### PR TITLE
planner: optimize the performance of `PlanCacheParamList.String()`

### LINKED ISSUE #46579: high cpu usage on PlanCacheParamList when using prepared statement
## Enhancement

After enable the prepared statement at client side, at the tidb cpu profile, we see 26% cpu on  variable.(*PlanCacheParamList).String for below workload, need cpu optimization on the function PlanCacheParamList

Workload: 50% write and 50% read
avg row size: 10KB
schema
```
create table test (`k` VARBINARY(64) NOT NULL, `v` BLOB, PRIMARY KEY (`k`) CLUSTERED);

```

f03f10b2-374e-414a-b242-1976d52a1476

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #46579

Problem Summary: planner: optimize the performance of `PlanCacheParamList.String()`

### What is changed and how it works?

planner: optimize the performance of `PlanCacheParamList.String()`

```
Before
  BenchmarkDatumsToString-8                 676939              1605 ns/op             952 B/op         37 allocs/op
  BenchmarkDatumsToStringStr-8              315628              3564 ns/op            1184 B/op          4 allocs/op
  BenchmarkDatumsToStringLongStr-8           17454             68606 ns/op           21794 B/op          4 allocs/op

After
  BenchmarkDatumsToString-8                1000000              1044 ns/op             792 B/op         34 allocs/op
  BenchmarkDatumsToStringStr-8            10546382               113.5 ns/op           584 B/op          2 allocs/op
  BenchmarkDatumsToStringLongStr-8         3918226               300.8 ns/op          2320 B/op          3 allocs/op

```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -2283,39 +2284,63 @@ func (ds *datumsSorter) Swap(i, j int) {
 	ds.datums[i], ds.datums[j] = ds.datums[j], ds.datums[i]
 }
 
+var bufPool = &sync.Pool{New: func() interface{} { return new(strings.Builder) }}
```
Comment (hawkingrei) [inline types/datum.go:2287]: Great!

TiDB uses strings.Builder in many places. but we don't have a pool to reuse it. so i think we can create ```util/sbpool``` to reuse it.

Comment (qw4990) [reply types/datum.go]: OK, I created a new file `utils/pool.go` to put all pools...

**Review by hawkingrei — APPROVED** [review]

File: types/datum.go [hunk 2]
```diff
@@ -2283,39 +2284,63 @@ func (ds *datumsSorter) Swap(i, j int) {
...
+		originalLen := -1
+		if len(str) > logDatumLen {
+			originalLen = len(str)
+			str = str[:logDatumLen]
+		}
 		if datum.Kind() == KindString {
-			strs = append(strs, fmt.Sprintf("%q", str))
+			builder.WriteString(`"`)
+			builder.WriteString(str)
+			builder.WriteString(`"`)
```
Comment (time-and-fate) [inline types/datum.go:2330]: Need to remind that this behavior is different from `%q`.

**Review by time-and-fate — APPROVED** [review]

**Comment by hawkingrei** [comment]
It also fixes the #TestShowStatsLocked.

**Pull request merged**